### PR TITLE
Document deprecation of version-specific SSL/TLS methods

### LINKED ISSUE #8989: doc: manual page should explicitly declare all single protocol TLS methods deprecated
The [SSL_CTX_new(3)](https://github.com/openssl/openssl/blob/5d238a1032fee0e4759c2ed7fbd09cb9d7125a72/doc/man3/SSL_CTX_new.pod) manual page states that the generic methods `TLS_*method()` should preferably be used instead of the single protocol methods, like e.g. `TLSv1_2_*method()`. However, it does not deprecate all single protocol methods entirely, only the `SSLv23_*method()`s are deprecated.

```
NOTES
       The SSL_CTX object uses method as connection method.  The methods exist in a generic type (for client and server use), a server only type, and
       a client only type.  method can be of the following types:

       TLS_method(), TLS_server_method(), TLS_client_method()
           These are the general-purpose version-flexible SSL/TLS methods.  The actual protocol version used will be negotiated to the highest
           version mutually supported by the client and the server.  The supported protocols are SSLv3, TLSv1, TLSv1.1, TLSv1.2 and TLSv1.3.
           Applications should use these methods, and avoid the version-specific methods described below.

       SSLv23_method(), SSLv23_server_method(), SSLv23_client_method()
           Use of these functions is deprecated. They have been replaced with the above TLS_method(), TLS_server_method() and TLS_client_method()
           respectively. New code should use those functions instead.

```

However, it turns out that all single protocoll methods are marked deprecated, in particular also `TLSv1_2_*method()`:

https://github.com/openssl/openssl/blob/5d238a1032fee0e4759c2ed7fbd09cb9d7125a72/include/openssl/ssl.h#L1902-L1904

Is this intentional? If yes, this needs to be documented more clearly in the manual page.

_(Side note: personally, I find it a bit confusing for the user that the TLS 1.2 methods are deprecated...)_

Related to issue #8726  (see the issue title and the first error message in the description)

### PR DESCRIPTION
**Pull request by mspncp**
In commit 2b8fa1d56cd3 the version-specific SSL/TLS methods were
deprecated. This patch improves the documentation of that change
by stating the deprecation more prominently in the manual page
and explaining the reason for the deprecation.

Fixes #8989

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

File: doc/man3/SSL_CTX_new.pod [hunk 1]
```diff
@@ -81,41 +81,51 @@ an existing B<SSL_CTX> structure.
...
+I<version-specific> methods: while the general-purpose methods will automatically
+negotiate the highest version mutually supported by the client and the server,
+the version-specific methods only support a single protocol version and will
+fail if that version is not supported by the peer.
+
+A significant disadvantage of using the version-specific methods is that
+an application will not automatically upgrade to newer protocol versions as
+they are added to OpenSSL, without having to be modified and recompiled.
+For that reason, applications should use only the general-purpose methods.
+In fact, all version-specific methods have been deprecated.
```
Comment (levitte) [inline doc/man3/SSL_CTX_new.pod]: I disagree with moving this text up from the items they originally came from, and I also think that if you really want to move this text around, you should do so in a different PR, where the pros and cons can be discussed without getting in the way of what you obviously want to do with this PR, i.e. mark the version specific methods as deprecated.

**Comment by kroeckx** [comment]
Is this something we want to merge in 1.1.0 too?

**Comment by mspncp** [comment]
> Is this something we want to merge in 1.1.0 too?

Since 1.1.0 is in security-fix-only mode, I don't think this will be allowed by our policy.

File: doc/man3/SSL_CTX_new.pod (hunk 1 above)
Comment (mspncp) [reply doc/man3/SSL_CTX_new.pod]: Sigh... it looks like you are right and I attempted too much at once. The reason why I moved this part to the beginning was because I wanted to motivate why the version-specific functions were deprecated, and I wanted to do it only once in order not to repeat myself. But I think you are right, this change is wrong for two reasons:

- People will scroll (resp. search) forward straightaway to the function they intend to use and will be confused not to find the desired information right there, but a few lines above.
- People don't want to learn about the reasons why things were deprecated (at least not here in the manual page), they just want to learn quickly which functions they can use and how to use them.

So I'll revert most of this and only leave the deprecation notes.

File: doc/man3/SSL_CTX_new.pod [hunk 2]
```diff
@@ -94,28 +94,31 @@ The actual protocol version used will be negotiated to the highest version
...
 
-Use of these functions is deprecated. They have been replaced with the above
-TLS_method(), TLS_server_method() and TLS_client_method() respectively. New
-code should use those functions instead.
+These functions do not exist anymore, they have been renamed to
+TLS_method(), TLS_server_method() and TLS_client_method() respectively.
+Currently, the old function calls are renamed to the corresponding new
+ones by preprocessor macros, to ensure that existing code which uses the
+old function names still compiles. However, using the old function names
+is deprecated and new code should call the new functions instead.
```
Comment (mspncp) [inline doc/man3/SSL_CTX_new.pod]: Leave this paragraph, or shall I revert it, too?

**Comment by mspncp** [comment]
ping

**Comment by mspncp** [comment]
Merged, thanks!

f308fa2573 Document deprecation of version-specific SSL/TLS methods [master]
8f75443f99 Document deprecation of version-specific SSL/TLS methods [1.1.1]
